Wrappers/Python/cil/optimisation/functions/ApproximateGradientSumFunction.py

```diff
@@ -0,0 +1,237 @@
+import numbers
+from abc import ABC, abstractmethod
+import numpy as np
+
+
+class ApproximateGradientSumFunction(SumFunction, ABC):
+    r"""ApproximateGradientSumFunction represents the following sum 
+
+    .. math:: \sum_{i=0}^{n-1} f_{i} = (f_{0} + f_{2} + ... + f_{n-1})
+
+    where there are :math:`n` functions. This function class has two ways of calling gradient:
+        - `full_gradient` calculates the gradient of the sum :math:`\sum_{i=0}^{n-1} \nabla f_{i}`
+        - `gradient` calls an `approximate_gradient` function which may be less computationally expensive to calculate than the full gradient
+    
+    
+
+    This class is an abstract base class and therefore is not able to be used as is. It is designed to be sub-classed with different approximate gradient implementations. 
+
+    Parameters:
+    -----------
+    functions : `list`  of functions
+                A list of functions: :math:`[f_{0}, f_{2}, ..., f_{n-1}]`. Each function is assumed to be smooth with an implemented :func:`~Function.gradient` method. All functions must have the same domain. The number of functions (equivalently the length of the list) must be strictly greater than 1. 
+    sampler: An instance of a CIL Sampler class ( :meth:`~optimisation.utilities.sampler`) or of another class which has a `next` function implemented to output integers in :math:`{0,...,n-1}`.
+        This sampler is called each time `gradient` is called and sets the internal `function_num` passed to the `approximate_gradient` function.  Default is `Sampler.random_with_replacement(len(functions))`. 
+
+    Note
+    -----
+    We ensure that the approximate gradient is of a similar order of magnitude to the full gradient calculation. For example, in the `SGFunction` we approximate the full gradient by :math:`n\nabla f_i` for an index :math:`i` given by the sampler. 
+    The multiplication by `math:`n` is a choice to more easily allow comparisons between stochastic and non-stochastic methods and between stochastic methods with varying numbers of subsets. 
+
+    Note
+    -----
+    Each time `gradient` is called the class keeps track of which functions have been used to calculate the gradient. This may be useful for debugging or plotting after using this function in an iterative algorithm:  
+        - `data_passes_indices` is a list of lists. Each time `gradient` is called a list is appended with with the indices of the functions have been used to calculate the gradient.  
+        - `data_passes` is a list. Each time `gradient` is called an entry is appended with  the proportion of the data used when calculating the approximate gradient  since the class was initialised (a full gradient calculation would be 1 full data pass). Warning: if your functions do not contain an equal `amount` of data, for example your data was not partitioned into equal batches, then you must first use the `set_data_partition_weights" function for this to be accurate.   
```

Comment:
```suggestion
    Each time :code:`gradient` is called the class keeps track of which functions have been used to calculate the gradient. This may be useful for debugging or plotting after using this function in an iterative algorithm:  
        - :code:`data_passes_indices` is a list of lists. Each time :code:`gradient` is called a list is appended with with the indices of the functions have been used to calculate the gradient.  
        - :code:`data_passes` is a list. Each time :code:`gradient` is called an entry is appended with  the proportion of the data used when calculating the approximate gradient  since the class was initialised (a full gradient calculation would be 1 full data pass). Warning: if your functions do not contain an equal `amount` of data, for example your data was not partitioned into equal batches, then you must first use the :code:`set_data_partition_weights` function for this to be accurate.   
```

Does `data_passes` still exist?